common/src/parse.cpp

```diff
@@ -79,60 +82,93 @@ pcl::console::parse_argument (int argc, const char * const * argv, const char *
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, bool &val)
 {
-  int index = find_argument (argc, argv, str) + 1;
+  long int dummy;
+  const auto ret = parse_argument (argc, argv, str, dummy);
+  val = static_cast<bool> (dummy);
+  return ret;
+}
 
+////////////////////////////////////////////////////////////////////////////////
+namespace pcl
+{
+namespace console
+{
+template <class T, class V = T(*)(const char*, const char**)> int
+parse_generic (V convert_func, int argc, const char* const* argv, const char* str, T& val)
+{
+  char *endptr;
+  int index = find_argument (argc, argv, str) + 1;
+  errno = 0;
+  
   if (index > 0 && index < argc )
-    val = atoi (argv[index]) == 1;
+  {
+    val = convert_func (argv[index], &endptr);  // similar to strtol, strtod, strtof
+  }
+  // handle out-of-range, junk at the end and no conversion
+  if (errno == ERANGE || *endptr != '\0' || str == endptr)
+  {
+    return -1;
+  }
 
   return (index - 1);
 }
+}
+}
 
 ////////////////////////////////////////////////////////////////////////////////
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, double &val)
 {
-  int index = find_argument (argc, argv, str) + 1;
-
-  if (index > 0 && index < argc )
-    val = atof (argv[index]);
-
-  return (index - 1);
+  return parse_generic(strtod, argc, argv, str, val);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, float &val)
 {
-  int index = find_argument (argc, argv, str) + 1;
-
-  if (index > 0 && index < argc )
-    val = static_cast<float> (atof (argv[index]));
-
-  return (index - 1);
+  return parse_generic(strtof, argc, argv, str, val);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, int &val)
 {
-  int index = find_argument (argc, argv, str) + 1;
+  long int dummy;
+  const auto ret = parse_argument (argc, argv, str, dummy);
+  if ((ret != -1) &&  (dummy >= std::numeric_limits<int>::min ()) && (dummy <= std::numeric_limits<int>::max ()))
+  {
+    val = static_cast<unsigned int> (dummy);
+  } else
+  {
+    return -1;
+  }
 
-  if (index > 0 && index < argc )
-    val = atoi (argv[index]);
+  return ret;
```

Comment:
It's a matter of style, but things get cleaner in this scenario if you write it like
```c++
if ((ret == -1) || (dummy < std::numeric_limits<int>::min ()) || (dummy > std::numeric_limits<int>::max ()))
    return -1;

val = static_cast<unsigned int> (dummy);
return ret;
```